### PR TITLE
Add comprehensive causal and custom masking support

### PR DESCRIPTION
**Pull request by bghira**
Implements attention masking with:
- AttentionMaskType enum (.none, .causal, .custom)
- Metal kernel-level masking
- Integration across forward/backward passes for gradient correctness
- Tests & docs

### CODE REVIEW

**Comment by philipturner** [comment]
This is great enthusiasm and fits well with the code base's design, but I designed the codebase to be absolutely bare-bones minimal, so I don't have to maintain it after I was (paid to) create it. Masks like this are one of many features beyond the critical core, which belong in a client's specific modification of the main library. I do not plan to ever merge any PRs into MFA, beyond typo or bug fixes to the existing code.

Contact @liuliu if you are interested in getting something done, and again best admiration for the enthusiasm.

**Comment by bghira** [comment]
i was almost certain you wouldn't want it, but figured to shuttle it up anyway. i've also created Rust and Obj-C bindings for a C-based FFI that links directly into the Swift library with zero-copy buffers. thanks for all the great work and cleanly written examples.

**Comment by liuliu** [comment]
Would love to merge this in certain ways. I maintain the C++ rewrite at https://github.com/liuliu/ccv/tree/unstable/lib/nnc/mfa/v2 (would like to move that out into a separate repo, but haven't got time). It did more tunes for things such as multi-head attention, and don't need the Swift compiler.

**Comment by liuliu** [comment]
I think the main concern are:

1. the broadcast semantics (that is simple though, since the implementation is single-head attention, so broadcast semantics is only relevant to multi-head);
2. the loading and related sram requirement would mess up some performance;
3. Uneven launch for causal attention might be wasteful.

---

On a related note, I *think* you don't need to integrate backward pass for masking? Since `masked_attn_score = attn_score + mask`, the gradient can directly flow into `attn_score`, or am I missing anything?

Feel free to contact me to discuss further!

**Comment by bghira** [comment]
i have to sleep, but my work i've pushed thus far is available here: https://github.com/bghira/universal-metal-flash-attention

but you should see how slow pytorch SDPA is with causal masking - this is faster, unbelievably enough. i'm aware it's a naive approach, but i'd hoped by opening the PR there'd be some feedback to improve it. definitely open to talking more about it

**Comment by philipturner** [comment]
I think causal masking can generalize all the way to sparse FlashAttention. In an old version of the repo, I experimented with block sparsity. It should be surprisingly easy to implement, compared to all the heavy lifting to discover the optimal algorithms supporting all 3 passes (fwd, bwd dQ, bwd dK/dV) with reasonable generality and applicability to all GPU architectures.

The main issue is Apple banning SIMD-group async copy with macOS 26, effectively imposing forced obsolescence on the M1 and M2 chips. This is not a surprise, as Moore's Law recently died and semiconductor lithography stopped improving. Just look at AMD's recent lineup not even containing a successor to the AMD 7900 XTX. But there's always room for people to have fun learning & experimenting with the inner details of digital algorithms. That's what life is about, I guess!

**Comment by liuliu** [comment]
> i have to sleep, but my work i've pushed thus far is available here: https://github.com/bghira/universal-metal-flash-attention
> 
> but you should see how slow pytorch SDPA is with causal masking - this is faster, unbelievably enough. i'm aware it's a naive approach, but i'd hoped by opening the PR there'd be some feedback to improve it. definitely open to talking more about it

Great! Cannot find your email through the repo trick. Email me i@liuliu.me if you want to discuss further!

**Comment by philipturner** [comment]
> but you should see how slow pytorch SDPA is with causal masking

I feel pleased to see the few people who recognize systemic problems like this. Extremely complex software systems are built with ten-thousand-dollar datacenter GPUs in mind, but the average person's experience is not designed for...
